Tests

ShastkivRuslan opened the pull request:
Added tests for all important methods;
![Screenshot 2024-12-04 152456](https://github.com/user-attachments/assets/614d1ab1-c45b-49a6-bd44-e76b895bf0e8)

ShastkivRuslan committed:
-transfer

ShastkivRuslan committed:
-transfer v2

ShastkivRuslan committed:
added tests to all controllers;

ShastkivRuslan committed:
added tests to service layer;

ShastkivRuslan committed:
tests -> done;
refactoring ->

ShastkivRuslan committed:
refactoring -> done;

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/OrderControllerTest.java:
```diff
@@ -0,0 +1,226 @@
+
+        List<OrderDto> actualDtos = getOrderDtosFromMvcResult(result, objectMapper);
+        assertEquals(expectedDtos.size(), actualDtos.size());
+
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/controller/OrderControllerTest.java:
```diff
@@ -0,0 +1,226 @@
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_CLASS
+)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class OrderControllerTest {
```
it would be great to add negative test cases. For example, try to place an order for an empty shopping cart

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/repository/OrderItemRepositoryTest.java:
```diff
@@ -0,0 +1,113 @@
+            """)
+    public void findAllByOrderId_validOrderId_returnPageableOrderItems() {
+        Page<OrderItem> actualPage = orderItemRepository.findAllByOrderId(ORDER_ID_1, PAGEABLE);
+        assertTrue(actualPage.hasContent());
```
Should you check the content?

JJJazl replied:
Let's check not only whether we have the content, but also what's inside)

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/service/OrderServiceTest.java:
```diff
@@ -0,0 +1,183 @@
+    private OrderMapper orderMapper;
+    @Mock
+    private OrderItemMapper orderItemMapper;
+    @Mock
```
I recommend you use Spy and simplify your test cases

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/service/OrderServiceTest.java:
```diff
@@ -0,0 +1,183 @@
+
+        when(userService.getUserId(authentication)).thenReturn(USER_ID);
+        when(shoppingCartService.findShoppingCart(USER_ID)).thenReturn(cart);
+        when(orderRepository.save(any(Order.class))).thenReturn(order);//todo
```
```suggestion
        when(orderRepository.save(any(Order.class))).thenReturn(order);
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/service/OrderServiceTest.java:
```diff
@@ -0,0 +1,183 @@
+            getAllOrdersByUserId()
+            - Should get all orders by user ID
+            """)
+    public void getAllOrdersByUserId_validRequest_returnPageableOrderDto() {
+        OrderItemDto orderItemDto = createOrderItemDto(ORDER_ITEM_ID_1);
+        OrderDto orderDto = createOrderDto(ORDER_ID_1, List.of(orderItemDto), Order.Status.PENDING);
+        Order order = createOrder(createUser(USER_ID), new ArrayList<>());
+        OrderItem orderItem = createOrderItem(order, FIRST_BOOK_ID);
+        order.getOrderItems().add(orderItem);
+        Page<Order> orderPage = new PageImpl<>(List.of(order), PAGEABLE, PAGE_SIZE_1);
+        PageImpl<OrderDto> expectedPage = new PageImpl<>(List.of(orderDto), PAGEABLE, PAGE_SIZE_1);
+
+        when(orderMapper.toDto(any(Order.class))).thenReturn(orderDto);
+        when(orderRepository.findAllByUserId(USER_ID, PAGEABLE)).thenReturn(orderPage);
+        Page<OrderDto> actualPage = orderService.getAllOrdersByUserId(USER_ID, PAGEABLE);
+
+        Assertions.assertEquals(expectedPage, actualPage);
```
```suggestion
    public void getAllOrdersByUserId_validRequest_returnPageableOrderDto() {
        // Given
        OrderItemDto orderItemDto = createOrderItemDto(ORDER_ITEM_ID_1);
        OrderDto orderDto = createOrderDto(ORDER_ID_1, List.of(orderItemDto), Order.Status.PENDING);
        Order order = createOrder(createUser(USER_ID), new ArrayList<>());
        OrderItem orderItem = createOrderItem(order, FIRST_BOOK_ID);
        order.getOrderItems().add(orderItem);
        Page<Order> orderPage = new PageImpl<>(List.of(order), PAGEABLE, PAGE_SIZE_1);
        PageImpl<OrderDto> expectedPage = new PageImpl<>(List.of(orderDto), PAGEABLE, PAGE_SIZE_1);

        when(orderMapper.toDto(any(Order.class))).thenReturn(orderDto);
        when(orderRepository.findAllByUserId(USER_ID, PAGEABLE)).thenReturn(orderPage);
        
        // When
        Page<OrderDto> actualPage = orderService.getAllOrdersByUserId(USER_ID, PAGEABLE);
        
        // Then
        Assertions.assertEquals(expectedPage, actualPage);
```
You can increase readability in this way

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/utils/OrderTestUtils.java:
```diff
@@ -0,0 +1,113 @@
+import ruslan.shastkiv.bookstore.model.Order;
+import ruslan.shastkiv.bookstore.model.OrderItem;
+import ruslan.shastkiv.bookstore.model.User;
+/*
+NOTE!!!
+in DB we have 2 test users with id 2 and 3
+user with id 3 have 2 orders with id 1 and 2
+order with id 1 have 2 order items with id 1 and 2 and books with id 1, 2
+order with id 2 have 1 order item with id 3 and book with id 3
+ */
```
better remove it for now

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/utils/OrderTestUtils.java:
```diff
@@ -0,0 +1,113 @@
+        orderItem.setQuantity(book.getId().intValue());
+        orderItem.setPrice(book.getPrice().multiply(BigDecimal.valueOf(orderItem.getQuantity())));
+        return orderItem;
+
```
```suggestion
```

liliia-ponomarenko commented on src/test/java/ruslan/shastkiv/bookstore/utils/ShoppingCartTestUtils.java:
```diff
@@ -0,0 +1,80 @@
+
+        return new ShoppingCartDto(userId, Set.of(cartItemDto));
+    }
+
```
```suggestion
```

liliia-ponomarenko commented on src/test/resources/scripts/cart/insert_cart_items.sql:
```diff
@@ -0,0 +1,3 @@
+INSERT INTO cart_items(id, shopping_cart_id, book_id, quantity)
+VALUES
+    (3, 3, 3, 3);
```
add one empty line at the end of the file  (check other files)

liliia-ponomarenko commented on src/test/resources/scripts/order/insert_order.sql:
```diff
@@ -0,0 +1,3 @@
+INSERT INTO orders (id, user_id, status, total, order_date, shipping_address)
+VALUES
+    (1, 3, 'PENDING', 9, NOW(), 'Shipping_address_3');
```
For tests better use expected data  (about NOW())

liliia-ponomarenko reviewed (changes requested):
You covered the code well with tests, that's a cool job!
I understand that you are limited by the program on the platform, but on a real project, it's better not to create such large PRs. Your colleagues won't appreciate it(

ShastkivRuslan committed:
refactored OrderTestUtils;
added javadoc for methods in OrderTestUtils;
added spy to order service tests for mocking behavior

ShastkivRuslan committed:
requiried changes was passed

ShastkivRuslan committed:
bugfix;

ShastkivRuslan commented:
- Went through all the scripts and added empty lines at the end where needed.
- Added JavaDoc to utility methods with more complex logic.
- Started using spy to simplify test writing.
- Added a few test cases for some methods.
- Refactored some utility methods.

sarakhmen commented on src/test/java/ruslan/shastkiv/bookstore/utils/OrderTestUtils.java:
```diff
@@ -34,80 +28,219 @@ public class OrderTestUtils {
-                objectMapper.readTree(result.getResponse().getContentAsString())
-                        .get("content"), new TypeReference<List<OrderDto>>() {}
-        );
+    /**
+     * Creates an instance of {@link UpdateOrderStatusRequestDto} with the specified order status.
+     *
+     * @param status the new status to set for the order;
+     * @return an {@link UpdateOrderStatusRequestDto} containing the specified status.
+     */
+
```
Do you really need all of these javadoc comments? As a best practice currently there is the next approach:
if you write your own API (not expose it to the other devs/companies to use, meaning your API isn't public) it's better to write the code in such a way there is no need to comment it. Excessive comments could be hard to maintain and require extra time to write it

sarakhmen replied:
if you do decide to have javadoc, then please remove extra empty line after comment (before method declaration)

sarakhmen commented on src/test/java/ruslan/shastkiv/bookstore/service/BookServiceTest.java:
```diff
@@ -70,12 +76,10 @@ void createBook_ValidCreateBookDto_ReturnBookDto() {
-        when(bookRepository.save(book)).thenReturn(book);
-        when(bookMapper.toDto(book)).thenReturn(expectedBookDto);
-
+        when(bookRepository.save(any(Book.class))).thenReturn(book);
```
As a best practice, I recommend not any() argument matches, it's better to use exact values when stubbing. The only case when any() is a good tool for you - is when there is no way to use exact matching :)

sarakhmen commented on src/test/java/ruslan/shastkiv/bookstore/service/OrderServiceTest.java:
```diff
@@ -82,16 +87,53 @@ public void placeOrderByUserId_validRequest_returnOrderDto() {
+            placeOrderByUserId()
+            - Should throw an exception when shopping cart is empty
+            """)
+    public void placeOrderByUserId_emptyCart_trowsException() {
```
please read method name once again and try to resolve a typo

sarakhmen commented on src/test/java/ruslan/shastkiv/bookstore/service/ShoppingCartServiceTest.java:
```diff
@@ -87,7 +96,7 @@ public void getShoppingCart_validId_returnShoppingCartDto() {
 
         when(shoppingCartRepository.findById(USER_ID)).thenReturn(Optional.of(cart));
-        when(shoppingCartMapper.toDto(cart)).thenReturn(shoppingCartDto);
+        //when(shoppingCartMapper.toDto(cart)).thenReturn(shoppingCartDto);
```
Do you need this commented code to be present here? Fix the same issue in all places

JJJazl replied:
Not fixed

sarakhmen reviewed (changes requested):
let's improve your solution :)

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/controller/OrderControllerTest.java:
```diff
@@ -0,0 +1,231 @@
+    )
+    public void placeOrder_validRequestDto_returnOrderDto() throws Exception {
+        User user = createUser(USER_ID);
+        Authentication authentication = getAuthentication(user);
```
Do you need it if you use `@WithMockUser`?
Please, check everywhere

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/controller/OrderControllerTest.java:
```diff
@@ -0,0 +1,231 @@
+
+        assertTrue(expectedDtos.stream()
+                .allMatch(expectedDto -> actualDtos.stream()
+                        .anyMatch(actualDto -> reflectionEquals(expectedDto, actualDto))));
```
You have only 3 fields in `OrderItemDto`, but so complicated logic to compare if expectedDtos and actualDtos are equal. It might look very 'functional', but indeed it's realy hard to read and debug. 
Please, try to keep it as simple as possible. Check it everywhere

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/repository/CartItemRepositoryTest.java:
```diff
@@ -0,0 +1,132 @@
+                actual.get().getBook().getTitle(),
+                CUSTOM_BOOK_TITLE.formatted(THIRD_BOOK_ID)
+        );
+        assertEquals(USER_ID, actual.get().getShoppingCart().getId());
```
Let's just create a few more variable to make it more readily and easy to understand)

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/repository/CartItemRepositoryTest.java:
```diff
@@ -0,0 +1,132 @@
+        Optional<CartItem> actual
+                = cartItemRepository.findByBookIdAndShoppingCartId(THIRD_BOOK_ID, USER_ID);
+
+        assertTrue(actual.isPresent());
+        assertEquals(
+                actual.get().getBook().getTitle(),
+                CUSTOM_BOOK_TITLE.formatted(THIRD_BOOK_ID)
+        );
+        assertEquals(USER_ID, actual.get().getShoppingCart().getId());
```
Same here

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/repository/RoleRepositoryTest.java:
```diff
@@ -0,0 +1,32 @@
+        Optional<Role> actualRole = roleRepository.findByRoleName(Role.RoleName.ROLE_USER);
+
+        assertTrue(actualRole.isPresent());
+        assertEquals(Role.RoleName.ROLE_USER, actualRole.get().getRoleName());
```
According to the javadoc, better way to get the content of an Optional is to use the `orElseThrow()` method.

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/service/OrderServiceTest.java:
```diff
@@ -0,0 +1,211 @@
+            placeOrderByUserId()
+            - Should throw an exception when shopping cart is empty
+            """)
+    public void placeOrderByUserId_emptyCart_trowsException() {
```
```suggestion
    public void placeOrderByUserId_emptyCart_throwsException() {
```

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/service/ShoppingCartServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+    public void findShoppingCart_invalidRequest_throwException() {
+        when(shoppingCartRepository.findById(INVALID_USER_ID)).thenReturn(Optional.empty());
+
+        assertThrows(EntityNotFoundException.class,
```
Check error message content

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/utils/OrderTestUtils.java:
```diff
@@ -0,0 +1,248 @@
+     * @return an {@link Order} instance with the specified user and populated items.
+     */
+
+    public static Order createOrderWithItems(User user, List<Long> bookIds) {
```
This should be a simple method that simply returns you an instance of the `Order` model with only the data you need. Don't complicate things

JJJazl commented on src/test/java/ruslan/shastkiv/bookstore/controller/AuthenticationControllerTest.java:
```diff
@@ -0,0 +1,116 @@
+
+        UserDto actualDto = objectMapper.readValue(
+                result.getResponse().getContentAsString(), UserDto.class);
+        Assertions.assertEquals(expectedDto, actualDto);
```
Use static import for `Assertions` methods

ShastkivRuslan committed:
changes after mentors review;

ShastkivRuslan commented:
Refactor test cases for readability and simplicity:
- Removed unnecessary use of authentication where not needed.
- Simplified DTO comparisons in `OrderControllerTest` for better readability.
- Added descriptive variables in `CartItemRepositoryTest` to improve clarity.
- Verified content in repository tests, ensuring expected data is checked.
- Renamed test methods for consistency (`throwsException` typo fix).
- Removed unused commented-out code in `ShoppingCartServiceTest`.
- Enhanced error message validation in exception handling tests.
- Applied static imports for assertion methods for cleaner code.
